File: electives/JPP/L2/GF.cpp
//
// Created by jaskow on 06.04.24.
//

#include <stdexcept>
#include "GF.hpp"
#include "lib.h"
void GF::checkP(const GF& L, const GF& R) {
    if(L.p != R.p)
        throw std::invalid_argument("[GF]ERROR: not matching p!");
}

GF::GF(const uint32_t& _p, const int64_t& _val) {
    if(_p == 0) {
        throw std::invalid_argument("[GF]ERROR: p must be > 0");
    }
    p = _p;
    setVal(_val);
}

uint32_t GF::getP() const noexcept{
    return p;
}
void GF::setP(uint32_t _p) noexcept{
    p = _p;
    this->setVal(val);
}

uint32_t GF::getVal() const noexcept{
    return val;
}
void GF::setVal(int64_t _val) noexcept{
    if(_val < 0 ) {
        val = (p + _val) % p;
    }
    else {
        val = _val;
    }
}

// compare
bool GF::operator==(const GF &R) const {
    checkP(*this, R);
    return this->val == R.val;
}
bool GF::operator!=(const GF &R) const {
    return !(*this == R);
}
bool GF::operator<(const GF &R) const {
    checkP(*this, R);
    return this->val < R.val;
}
bool GF::operator>(const GF &R) const {
    return R < *this;
}
bool GF::operator<=(const GF &R) const {
    return !(*this > R);
}
bool GF::operator>=(const GF &R) const {
    return !(*this < R);
}

// arithmetic
GF GF::operator+() const noexcept {
    return *this;
}
GF GF::operator-() const noexcept{
    return GF(p, p - val);
}
GF GF::operator-(const GF &R) const {
    checkP(*this, R);
    return ((-R) + (*this));
}
GF GF::operator+(const GF &R) const {
    checkP(*this, R);
    uint32_t _val = (val + R.val) % p;
    return GF(R.p, _val);
}
GF GF::operator*(const GF &R) const {
    checkP(*this, R);
    uint32_t _val = (val * R.val) % p;
    return GF(p, _val);
}
GF GF::operator/(const GF &R) const {
    checkP(*this, R);

    uint32_t RValReverse = ILDES((uint64_t)R.val, (uint64_t)R.p, RGCD(R.val, R.p)).x;

    uint32_t _val = (val * RValReverse) % p;
    return GF(p, _val);
}

// assignment
GF& GF::operator+=(const GF &other) {
    *this = ((*this) + other);
    return *this;
}
GF& GF::operator-=(const GF &other) {
    *this = (*this - other);
    return *this;
}
GF& GF::operator*=(const GF &other) {
    *this = ((*this) * other);
    return *this;
}
GF& GF::operator/=(const GF &other) {
    *this = (*this / other);
    return *this;
}

std::ostream &operator<<(std::ostream &out, const GF &gf) {
    out << "p   = " << gf.p
        << "val = " << gf.val;
    return out;
}

File: electives/JPP/L2/java/src/main/java/org/example/GF.java
package org.example;

public class GF {
    private int p;
    private long val;

    public GF(int _p, long _val) {
        if(_p == 0) {
            throw new IllegalArgumentException("[GF]ERROR: p must be > 0");
        }
        this.p = _p;
        this.setVal(_val);
    }

    public int getP() {
        return this.p;
    }

    public void setP(int _p) {
        this.p = _p;
        this.setVal(this.val);
    }

    public long getVal() {
        return this.val;
    }

    public void setVal(long _val) {
        if(_val < 0 ) {
            this.val = (this.p + _val) % this.p;
        }
        else {
            this.val = _val;
        }
    }

    private static void checkP(GF L, GF R) {
        if(L.p != R.p)
            throw new IllegalArgumentException("[GF]ERROR: not matching p!");
    }

    // compare
    public boolean equals(GF R) {
        checkP(this, R);
        return this.val == R.val;
    }

    public boolean notEquals(GF R) {
        return !this.equals(R);
    }

    public boolean lessThan(GF R) {
        checkP(this, R);
        return this.val < R.val;
    }

    public boolean greaterThan(GF R) {
        return R.lessThan(this);
    }

    public boolean lessThanOrEqual(GF R) {
        return !this.greaterThan(R);
    }

    public boolean greaterThanOrEqual(GF R) {
        return !this.lessThan(R);
    }

    // arithmetic
    public GF plus() {
        return this;
    }

    public GF minus() {
        return new GF(this.p, (this.p - this.val) % p);
    }

    public GF subtract(GF R) {
        checkP(this, R);
        return this.minus().add(R);
    }

    public GF add(GF R) {
        checkP(this, R);
        long _val = (this.val + R.val) % this.p;
        return new GF(R.p, _val);
    }

    public GF multiply(GF R) {
        checkP(this, R);
        long _val = (this.val * R.val) % this.p;
        return new GF(this.p, _val);
    }

    public GF divide(GF R) {
        checkP(this, R);
        long RValReverse = Lib.iLDES(R.val, R.p, Lib.iGCD(R.val, R.p))[0];
        long _val = (this.val * RValReverse) % this.p;
        return new GF(this.p, _val);
    }

    // assignment
    // TODO: fix this
    public GF plusEquals(GF other) {
        this.setVal();
        return this;
    }
    // TODO: fix this
    public GF minusEquals(GF other) {
//        this.setVal((this.subtract(other)).getVal());
        return this;
    }

    public GF multiplyEquals(GF other) {
        this.setVal((this.multiply(other)).getVal());
        return this;
    }

    public GF divideEquals(GF other) {
        this.setVal((this.divide(other)).getVal());
        return this;
    }

    @Override
    public String toString() {
        return "p   = " + this.p + "\nval = " + this.val;
    }
}